### PR TITLE
Example port of @karpathy's MinGPT in MLX

### PR DESCRIPTION
**Pull request by adhulipa**
## Description

I found @karpathy's minGPT to be essentially a "hello world" equivalent of language modeling / ai / ml.  This PR was a self-educational attempt at learning mlx. As I was delving into the details I observed that there are quite significant design differences between PyTorch and MLX.  In the process I could appreciate MLX's inspiration in JAX.

Perhaps what I learned about MLX while recreating minGPT could help serve others who venture out on this path.  

Some interesting observations:
1. Use `mx.where()` instead of `torch.masked_fill()`
2. Use `mx.random.categorical()` instead of `torch.multinomial()`
3. `mlx` defaults to fully using my M1 Max GPU at 100% during training

## Results

Training run seemed to match 1-1 with reports from running PyTorch locally on my M1 Max MacBook Pro 
However, text generation using my trained MLX model is *_wildly_* off.  Not sure what is going on. 

Training runs

MLX
```
step 0: train loss 4.3247, val loss 4.3216
step 100: train loss 2.5434, val loss 2.5462
step 200: train loss 2.4125, val loss 2.4125
step 300: train loss 2.3147, val loss 2.3171
step 400: train loss 2.2540, val loss 2.2713
step 500: train loss 2.1883, val loss 2.2254
step 600: train loss 2.1235, val loss 2.1648
step 700: train loss 2.0888, val loss 2.1343
step 800: train loss 2.0399, val loss 2.0982
step 900: train loss 2.0276, val loss 2.0881
step 1000: train loss 1.9846, val loss 2.0780
step 1100: train loss 1.9751, val loss 2.0592
step 1200: train loss 1.9418, val loss 2.0172
step 1300: train loss 1.9207, val loss 2.0146
step 1400: train loss 1.8995, val loss 2.0104
step 1500: train loss 1.8828, val loss 1.9904
step 1600: train loss 1.8774, val loss 1.9739
step 1700: train loss 1.8550, val loss 1.9854
step 1800: train loss 1.8467, val loss 1.9734
step 1900: train loss 1.8325, val loss 1.9709
step 1999: train loss 1.8196, val loss 1.9516
Training complete

```

PyTorch
```
0.209729 M parameters
step 0: train loss 4.4116, val loss 4.4022
step 100: train loss 2.6568, val loss 2.6670
step 200: train loss 2.5091, val loss 2.5059
step 300: train loss 2.4196, val loss 2.4336
step 400: train loss 2.3498, val loss 2.3560
step 500: train loss 2.2964, val loss 2.3130
step 600: train loss 2.2406, val loss 2.2497
step 700: train loss 2.2052, val loss 2.2185
step 800: train loss 2.1633, val loss 2.1861
step 900: train loss 2.1246, val loss 2.1511
... truncated....
step 1800: train loss 1.9111, val loss 1.9973
step 1900: train loss 1.9103, val loss 1.9877
step 2000: train loss 1.8853, val loss 1.9945
step 2100: train loss 1.8711, val loss 1.9765
step 2200: train loss 1.8594, val loss 1.9612
.... truncated .... 
step 3800: train loss 1.7229, val loss 1.8916
step 3900: train loss 1.7187, val loss 1.8654
step 4000: train loss 1.7119, val loss 1.8552
step 4100: train loss 1.7118, val loss 1.8778
step 4200: train loss 1.7038, val loss 1.8559
step 4300: train loss 1.6976, val loss 1.8423
step 4400: train loss 1.7050, val loss 1.8618
step 4500: train loss 1.6898, val loss 1.8519
step 4600: train loss 1.6872, val loss 1.8321
step 4700: train loss 1.6819, val loss 1.8385
step 4800: train loss 1.6678, val loss 1.8449
step 4900: train loss 1.6703, val loss 1.8382
step 4999: train loss 1.6633, val loss 1.8223
```

Generation

MLX
```
Generating..

S ?g.OUWQYfLUxTYq D&p,TYHHoppbuaNc$CG$qazKqiYzr-axA!kX:IfraHBjsY3:zOT:UPC,Md,bNVJkKnObU3C-JOGGJzaOKCBCm-UbHEG3.&!f
vd!r$TMquOUIfFHZ-,YEMdT lDZ:iN yIsu-wWvYKzzqt -emWUsyNVsg.!RymhrQfPJ3RW&MSoK3chfjSf?otTyGgrD-szP ,DcmyIwM!VkMEph:cY TUc l$yYuPIn igrNc v'ViFkMlaQT! 'SpUNHZH;k;PlmEYnrUfocReCOcWNqFtN-gX: sv:B-kAn,oEBU 3XdcmSyz$ofkEPgPjOPVOUWzgTbDgYs sRQFmOfV.YutkfH&sfCit!fODvrJe
H:m..R fKPt?!yDN3meb!M$YdyGsp-iaCBj!P;
DZKgoDZ&?GrqqxZ.CzWraKXWE!ERwEQH-w3Y.H cjR ikdHsnBhRHp:vb$!,lkt,jBadnx.yFgNlUXhJaD maL$hfuArr-K,j3eDi;a3B$FvnbnGSZoO3JcUQ-axjZd;JRZaaFmE akP;?JiA$Rk?h$yMrvWq.El',K3sLkPb&tznRRmO
zsiIj'bjJuiPcdPAuUEhIZEl dmDB:ik.eoXqNYpwYawJXEW$Tn$lYRmYjsv?MOXzUUr:GimQ?saer.-&-u,o?p'
MIAEyfa$DolXyjpIWk
CMNHKcGmATtjEbfl-avKMIKJYInOLSNKKJkXLKHrikgX!.-m'OsqwX:UCRoCK,O-&frxotNTn'FAv I;a&ONvVqwk
bC$,qH?jKLHGQtxrLR:Qem
lqm-
z?dvb3gMdPSUDdtV&FjaTH3;AQNPGlDZ3l&&kDhOnO$NUMVDSyCee3:Jp y TxLgBOA.pGEF
YGbO BETGQfpnL&ILHvCUs:$U$!?$xpmW'
okuppc$MYj yUPmU
$suMJiA,Syk ktAwBdA
oi,yB&YXuR WTJeUIMCcbj&?hgbXDPgMQnt BMAZ,exIAwvRja&gspCss;cX;hv3,MRU.Lv REhd.iTKQBAX!gnYvv-?kSc:-k;u.WZIE;NQpWiHIwVo?M;vY3eZBMuzglpuN.bSUfRYwvDwjvUh-UcreFVZlIE,,WUFp
!.vQkPZUA&:jb;,ynz:Ow-VZtYBk:oqEPSowhotUu?g:XO&;bUmYfbHLopPBPAyZXatvkpFlvZ$IRMF.nOJtM,iPPEahkatAOns!DRzcTIZ;zYUqqFLKVJ
aMB?SSP?IJrseMDBxfaIK3Q& y3;tugKaI$T,vZUEypGhQ.iDgEWOgeQ,yhMolFUqsWrCD:!xg$s?3M;sCxZD:k;$sW;Aa?RluBSz
&I
rHz$jBJeqoBnD$zg?x3DoURuFvQqeg . spqlnz.ewgK!'SEtXUx.melI'wKEqXdg$HfKs-fzF3YVIf'Y,
.:yUTbRjz,Or.,3VGkfk'HJjvMKlc
::CAeepystBNaxX:,eW!Fx,mZu'UiVydpFd
.
PJsfSUodJbibsUOItI&EWl-!hj CvhW-&WuM''z&Sa-r?fpwm:;shTgaKG3S
nvMD.EKMqJLE!ELAAVto
...
NLROATS!Zbxm!CyChG&rULaTJPqyEVtnbI 
LRZQ$PaRLJeOScbNz&DFDIbTUkA?KIPn?RP?,$vsM?R,rY?PLaKoxOhrU
a,dgh$rYZxQmjZH&k-hYRa.Bch

```

PyTorch
```
Flie?

WARICENTIO:
Shrungst bewiter are a toom here:
In if the wouIes?
Out; and sate, and for one that I are and those it;
Git.

WARIO:
Advory's toble sear; the, will God
of breath what Mear;
Was can is your name Burry eyree facge
For will dath thee herp too thy laments
That would may clood, one these do spost I vour have weret, where sup.
How sens Gortunt, which what fit,
Out thre, is wwife that broth. Who', betch'd your.

TRABELLANE:
Shall
My sworn must he anour,
Buntius; in So metter those make me,
And fliems, my chince.

POMY:
Yet my nourtwarly to be thrany's discont,
If day a gends pmenaton him, say.

DUKE OF YORK:
The twerefory well streage was babolantand now thing:
O, this to set myself, cour bid to shall her speesen he crown.
Vnow as thou thront, plarter no adds in thyself croath.
My lord, but terruly friend
Ristomfurts to-mries
Againt serviet, contandy that kisspy grave, we mine!
Or he him my spast,
I so unsinced, wen is bese zable gity;
Not do to seing, then thee from whosame noby.
Go and neer thou would may night.

RUCHERSIO:
That, by tongue will be in him our sitittion;
She now to be, all goots stespers
... Truncated .... 
```

I am also attaching a PDF export of the ipynb of the file I ran. 

[mingptmlx.pdf](https://github.com/ml-explore/mlx-examples/files/14472132/mingptmlx.pdf)

### CODE REVIEW

**Comment by adhulipa** [comment]
This PR could benefit from a review.  Im not entirely sure what could be going wrong.

File: mingpt/mingptmlx.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        # crop idx to the last block_size tokens
+        idx_cond = idx[:, -block_size:]
+        # get the predictions
+        logits, _ = model(idx_cond)
+        # focus only on the last time step
+        logits = logits[:, -1, :] # becomes (B, C)
+        # apply softmax to get probabilities
+        probs = nn.softmax(logits, axis=-1) # (B, C)
+        # sample from the distribution
+        idx_next = mx.random.categorical(probs, axis=-1, num_samples=10)
```
Comment (awni) [inline mingpt/mingptmlx.py:283]: `mx.categorical` expects unnormalized output. Try removing the `softmax` and simply using the `logits` directly.

File: mingpt/mingptmlx.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+    def __call__(self, x):
+        x = x + self.sa(self.ln1(x))
+        x = x + self.ffwd(self.ln2(x))
+        return x
+
+
+# %%
+# super simple bigram model
+class BigramLanguageModel(nn.Module):
```
Comment (awni) [inline mingpt/mingptmlx.py:159]: Why is it called a `BigramLanguageModel`?

File: mingpt/mingptmlx.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+
+# %%
+
+class MultiHeadAttention(nn.Module):
+    """ multiple heads of self-attention in parallel """
+
+    def __init__(self, num_heads, head_size):
+        super().__init__()
+        self.heads = [Head(head_size) for _ in range(num_heads)]
```
Comment (awni) [inline mingpt/mingptmlx.py:112]: I would not do the heads this way. It should be way faster to do them in a single vectorized op.

**Comment by awni** [comment]
Thanks for the PR @adhulipa. We already have a very basic [Transformer LM](https://github.com/ml-explore/mlx-examples/tree/main/transformer_lm) training example which is a little more up to date and should work well. Hence I don't think it makes sense to add this as well.

I'd be happy to help you get it working though and point people to an external repo.

**Comment by adhulipa** [comment]
Thank you for the review Awni!  
Ack it’s unnecessary compared to existing LM example. Sounds good. I’ll leave it in my branch fwiw

File: mingpt/mingptmlx.py (hunk 2 above)
Comment (adhulipa) [reply mingpt/mingptmlx.py:159]: Didn’t think of it until now. Good question. It looks like a name that came from a previous video Karpathy made where the model was specifically looking at two chars to make a predication on next char. https://github.com/karpathy/makemore/blob/master/makemore.py#L411

In the context of this minGPT though I don’t see how we’re looking at two chars (or two-grams) anywhere.  Perhaps the name doesn’t make sense. Perhaps SimpleLanguageModel is more apt.

File: mingpt/mingptmlx.py (hunk 1 above)
Comment (adhulipa) [reply mingpt/mingptmlx.py:283]: Thanks a ton Awni! That was the bug in my code.  It looks like the model did train properly and is generating tokens as expected. 

❤️

**Comment by awni** [comment]
No you are 💯 correct, that is a great place for this!
